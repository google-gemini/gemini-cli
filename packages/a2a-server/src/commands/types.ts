/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import type { Config } from '@google/gemini-cli-core';

export interface CommandArgument {
  readonly name: string;
  readonly description: string;
  readonly isRequired?: boolean;
}

export interface Command {
  readonly name: string;
  readonly description: string;
  readonly arguments?: CommandArgument[];
  readonly subCommands?: Command[];
  readonly topLevel?: boolean;

  execute(config: Config, args: string[]): Promise<CommandExecutionResponse>;
}

export interface CommandExecutionResponse {
  readonly name: string;
  readonly data: unknown;
}
