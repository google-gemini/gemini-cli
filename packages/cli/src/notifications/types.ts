/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Defines the structure for individual notification event settings.
 */
export interface NotificationEventSettings {
  enabled: boolean;
  sound: 'system' | 'custom';
  customPath?: string; // Path to custom sound file, only if sound is 'custom'
  timeout?: number; // Optional timeout for idle alerts
}

/**
 * Defines the structure for the overall notification configuration.
 */
export interface NotificationSettings {
  enabled: boolean;
  events: {
    inputRequired: NotificationEventSettings;
    taskComplete: NotificationEventSettings;
    idleAlert: NotificationEventSettings;
  };
}

/**
 * Defines the event types for notifications.
 */
export type NotificationEventType = 'inputRequired' | 'taskComplete' | 'idleAlert';
