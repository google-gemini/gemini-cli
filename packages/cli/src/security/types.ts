/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

// Shared types for MCP server security modules
export type GuardDecision = "allow" | "allow_sanitized" | "escalate" | "block";

export interface GuardResult {
  decision: GuardDecision;
  sanitized?: string;
  reasons: string[];
  tags: string[]; // e.g., ["system-override","tool-coercion","obfuscation"]
}

export interface ContextMeta {
  source: "user" | "rag" | "tool";
  userRole: "user" | "admin" | "service";
  toolAcl: string[];       // allowed tool names
  provenance?: string;     // URL/doc/tool name
  conversationId: string;  // for cross-turn auditing
}

export interface ServerOptions {
  port?: number;
  host?: string;
  policyPath?: string;
  testMode?: boolean; // enable fuzz harness
  enableSecurity?: boolean; // enable security middleware
  enableMetrics?: boolean; // enable real-time metrics
  metricsPort?: number; // port for metrics dashboard
}
