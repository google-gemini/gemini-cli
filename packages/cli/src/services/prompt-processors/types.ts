/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Defines the interface for a prompt processor, a module that can transform
 * a prompt string before it is sent to the model. Processors are chained
 * together to create a processing pipeline.
 */
export interface IPromptProcessor {
  /**
   * Processes a prompt string, applying a specific transformation.
   *
   * @param prompt The current state of the prompt string.
   * @param args The raw argument string from the user's command invocation.
   * @param fullCommand The full command the user typed (e.g., /git:commit foo).
   * @returns The transformed prompt string as a Promise.
   *
   * @todo Add `context: CommandContext` back when more complex processors
   * that require access to services or UI are needed.
   */
  process(prompt: string, args: string, fullCommand: string): Promise<string>;
}

/**
 * The placeholder string for shorthand argument injection in custom commands.
 */
export const SHORTHAND_ARGS_PLACEHOLDER = '{{args}}';
