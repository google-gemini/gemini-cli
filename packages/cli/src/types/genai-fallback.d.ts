// Minimal local fallbacks for @google/genai types when upstream types aren't available during build.
// These are type-only declarations to keep the consumer compiling. Replace with real upstream types when available.

declare module '@google/genai' {
  // Keep types permissive to avoid leaking runtime coupling in this repo build.
  export type Part = { text?: string; name?: string; type?: string } | string | any;
  export type PartListUnion = Part | Part[] | null;
  export type PartUnion = Part | PartListUnion;
  export type FunctionResponse = { name?: string; arguments?: any } | any;
  export type FunctionCall = { name?: string; args?: any } | any;

  export enum FinishReason {
    FINISH_REASON_UNSPECIFIED = 0,
    STOP = 1,
    MAX_TOKENS = 2,
    SAFETY = 3,
    RECITATION = 4,
    LANGUAGE = 5,
    BLOCKLIST = 6,
    PROHIBITED_CONTENT = 7,
    SPII = 8,
    OTHER = 9,
    MALFORMED_FUNCTION_CALL = 10,
    IMAGE_SAFETY = 11,
    UNEXPECTED_TOOL_CALL = 12,
  }

  export type Content = any;
  export type GenerateContentConfig = any;
}
