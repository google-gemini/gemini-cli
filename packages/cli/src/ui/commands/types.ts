/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { Config, GitService, Logger } from '@google/gemini-cli-core';
import { LoadedSettings } from '../../config/settings.js';
import { HistoryItem, HistoryItemWithoutId, Message } from '../types.js';
import { UseHistoryManagerReturn } from '../hooks/useHistoryManager.js';
import { SessionStatsState } from '../contexts/SessionContext.js';

// Grouped dependencies for clarity and easier mocking
export interface CommandContext {
  // Core services and configuration
  services: {
    config: Config | null;
    settings: LoadedSettings;
    git: GitService | undefined;
    logger: Logger;
  };
  // UI state and history management
  ui: {
    history: HistoryItem[];
    addItem: UseHistoryManagerReturn['addItem'];
    clearItems: UseHistoryManagerReturn['clearItems'];
    loadHistory: UseHistoryManagerReturn['loadHistory'];
    refreshStatic: () => void;
    setQuittingMessages: (messages: HistoryItem[]) => void;
    pendingHistoryItems: HistoryItemWithoutId[];
  };
  // Functions to open dialogs/modals
  dialogs: {
    openTheme: () => void;
    openAuth: () => void;
    openEditor: () => void;
    openPrivacy: () => void;
    setShowHelp: (show: boolean) => void;
  };
  // Specific actions that interact with other hooks/state
  actions: {
    performMemoryRefresh: () => Promise<void>;
    toggleCorgiMode: () => void;
    setPendingCompression: (item: HistoryItemWithoutId | null) => void;
  };
  // Session-specific data
  session: {
    stats: SessionStatsState;
  };
  // Low-level utilities
  utils: {
    onDebugMessage: (message: string) => void;
    addMessage: (message: Message) => void;
  };
}

export interface SlashCommandActionReturn {
  shouldScheduleTool?: boolean;
  toolName?: string;
  toolArgs?: Record<string, unknown>;
  message?: string; // For simple messages or errors
}

// The standardized contract for any command in the system.
export interface SlashCommand {
  name: string;
  altName?: string;
  description?: string;

  // The action to run. Optional for parent commands that only group sub-commands.
  action?: (
    context: CommandContext,
    args: string,
  ) =>
    | void
    | SlashCommandActionReturn
    | Promise<void | SlashCommandActionReturn>;

  // Provides argument completion (e.g., completing a tag for `/chat resume <tag>`).
  completion?: (
    context: CommandContext,
    partialArg: string,
  ) => Promise<string[]>;

  subCommands?: SlashCommand[];
}
