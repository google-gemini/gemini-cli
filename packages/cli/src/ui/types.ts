/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  ToolCallConfirmationDetails,
  ToolResultDisplay,
} from '@gemini-code/server';

// Only defining the state enum needed by the UI
export enum StreamingState {
  Idle = 'idle',
  Responding = 'responding',
  WaitingForConfirmation = 'waiting_for_confirmation',
}

// Copied from server/src/core/turn.ts for CLI usage
export enum GeminiEventType {
  Content = 'content',
  ToolCallRequest = 'tool_call_request',
  // Add other event types if the UI hook needs to handle them
}

export enum ToolCallStatus {
  Pending = 'Pending',
  Canceled = 'Canceled',
  Confirming = 'Confirming',
  Executing = 'Executing',
  Success = 'Success',
  Error = 'Error',
}

export interface ToolCallEvent {
  type: 'tool_call';
  status: ToolCallStatus;
  callId: string;
  name: string;
  args: Record<string, never>;
  resultDisplay: ToolResultDisplay | undefined;
  confirmationDetails: ToolCallConfirmationDetails | undefined;
}

export interface IndividualToolCallDisplay {
  callId: string;
  name: string;
  description: string;
  resultDisplay: ToolResultDisplay | undefined;
  status: ToolCallStatus;
  confirmationDetails: ToolCallConfirmationDetails | undefined;
}

export interface HistoryItemBase {
  text?: string; // Text content for user/gemini/info/error messages
}

// Using Omit<HistoryItem, 'id'> seems to have some issues with typescript's
// type inference e.g. historyItem.type === 'tool_group' isn't auto-inferring that
// 'tools' in historyItem.
export type HistoryItemWithoutId = HistoryItemBase &
  (
    | { type: 'user'; text: string }
    | { type: 'gemini'; text: string }
    | { type: 'gemini_content'; text: string }
    | { type: 'info'; text: string }
    | { type: 'error'; text: string }
    | { type: 'tool_group'; tools: IndividualToolCallDisplay[] }
  );

export type HistoryItem = HistoryItemWithoutId & { id: number };

// Message types used by internal command feedback (subset of HistoryItem types)
export enum MessageType {
  INFO = 'info',
  ERROR = 'error',
  USER = 'user',
  // Add GEMINI if needed by other commands
}

// Simplified message structure for internal feedback
export interface Message {
  type: MessageType;
  content: string; // Renamed from text for clarity in this context
  timestamp: Date; // For consistency, though addItem might use its own timestamping
}
