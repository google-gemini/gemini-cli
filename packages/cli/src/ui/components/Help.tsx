/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { Box, Text } from 'ink';
import { Colors } from '../colors.js';
import { SlashCommand } from '../hooks/slashCommandProcessor.js';

interface Help {
  commands: SlashCommand[];
}

export const Help: React.FC<Help> = ({ commands }) => (
  <Box flexDirection="column" marginBottom={1}>
    <Text bold color={Colors.Foreground}>
      Abilities:
    </Text>
    <Text color={Colors.Foreground}> * Use tools to read and write files</Text>
    <Text color={Colors.Foreground}>
      {' '}
      * Semantically search and explain code
    </Text>
    <Text color={Colors.Foreground}> * Execute bash commands</Text>
    <Box height={1} />
    <Text bold color={Colors.Foreground}>
      Commands:
    </Text>
    {commands.map((command: SlashCommand) => (
      <Text key={command.name} color={Colors.SubtleComment}>
        <Text bold color={Colors.AccentPurple}>
          {' '}
          /{command.name}
        </Text>
        {command.description && ' - ' + command.description}
      </Text>
    ))}
    <Text color={Colors.SubtleComment}>
      <Text bold color={Colors.AccentPurple}>
        {' '}
        !{' '}
      </Text>
      shell command
    </Text>
    <Text color={Colors.SubtleComment}>
      <Text bold color={Colors.AccentPurple}>
        {' '}
        ${' '}
      </Text>
      echo hello world
    </Text>
  </Box>
);
