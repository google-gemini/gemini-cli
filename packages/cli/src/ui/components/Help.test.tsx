/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/** @vitest-environment jsdom */

import { render } from 'ink-testing-library';
import { describe, it, expect, beforeAll } from 'vitest';
import { Help } from './Help.js';
import type { SlashCommand } from '../commands/types.js';
import { CommandKind } from '../commands/types.js';
import '../../i18n/index.js';

const mockCommands: readonly SlashCommand[] = [
  {
    name: 'test',
    description: 'A test command',
    kind: CommandKind.BUILT_IN,
  },
  {
    name: 'hidden',
    description: 'A hidden command',
    hidden: true,
    kind: CommandKind.BUILT_IN,
  },
  {
    name: 'parent',
    description: 'A parent command',
    kind: CommandKind.BUILT_IN,
    subCommands: [
      {
        name: 'visible-child',
        description: 'A visible child command',
        kind: CommandKind.BUILT_IN,
      },
      {
        name: 'hidden-child',
        description: 'A hidden child command',
        hidden: true,
        kind: CommandKind.BUILT_IN,
      },
    ],
  },
];

describe('Help Component', () => {
  beforeAll(async () => {
    // Ensure i18next is ready and has loaded resources
    const i18next = (await import('../../i18n/index.js')).default;
    // Wait for i18next to be fully initialized
    if (!i18next.isInitialized) {
      await i18next.init();
    }
  });

  it('should not render hidden commands', () => {
    const { lastFrame } = render(<Help commands={mockCommands} />);
    const output = lastFrame();

    expect(output).toContain('/test');
    expect(output).not.toContain('/hidden');
  });

  it('should not render hidden subcommands', () => {
    const { lastFrame } = render(<Help commands={mockCommands} />);
    const output = lastFrame();

    expect(output).toContain('visible-child');
    expect(output).not.toContain('hidden-child');
  });

  it('renders basic help content', () => {
    const { lastFrame } = render(<Help commands={mockCommands} />);
    const output = lastFrame();

    // Test i18n content
    expect(output).toContain('Basics:');
    expect(output).toContain('Commands:');
    expect(output).toContain('Keyboard Shortcuts:');
    expect(output).toContain('Add context: Use @ to specify files');
    expect(output).toContain('Shell mode: Execute shell commands via !');
    expect(output).toContain('shell command');
    expect(output).toContain('Model Context Protocol command');
  });

  it('renders platform-specific shortcuts correctly', () => {
    const { lastFrame } = render(<Help commands={mockCommands} />);
    const output = lastFrame();

    // Test platform-agnostic shortcuts
    expect(output).toContain('Ctrl+C - Quit application');
    expect(output).toContain('Ctrl+L - Clear the screen');
    expect(output).toContain('Ctrl+Y - Toggle YOLO mode');
    expect(output).toContain('Alt+Left/Right - Jump through words');

    // Test platform-specific shortcuts
    if (process.platform === 'win32') {
      expect(output).toContain('Ctrl+Enter - New line');
      expect(output).toContain('Ctrl+X - Open input in external editor');
    } else if (process.platform === 'darwin') {
      expect(output).toContain('Ctrl+J - New line');
      expect(output).toContain(
        'Ctrl+X / Meta+Enter - Open input in external editor',
      );
    } else {
      // Linux and other platforms
      expect(output).toContain(
        'Ctrl+J - New line (Alt+Enter works for certain linux distros)',
      );
      expect(output).toContain('Ctrl+X - Open input in external editor');
    }
  });
});
