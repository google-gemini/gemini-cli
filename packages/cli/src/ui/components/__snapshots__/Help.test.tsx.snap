// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Help Component > renders help component with mock commands (Linux only) 1`] = `
"╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                  │
│ Basics:                                                                                          │
│ Add context: Use @ to specify files for context (e.g., @src/myFile.ts) to target specific files  │
│ or folders.                                                                                      │
│ Shell mode: Execute shell commands via ! (e.g., !npm run start) or use natural language (e.g.    │
│ start server).                                                                                   │
│                                                                                                  │
│ Commands:                                                                                        │
│  /test - A test command                                                                          │
│  /parent - A parent command                                                                      │
│    visible-child - A visible child command                                                       │
│  ! - shell command                                                                               │
│ [MCP] - Model Context Protocol command (from external servers)                                   │
│                                                                                                  │
│ Keyboard Shortcuts:                                                                              │
│ Alt+Left/Right - Jump through words in the input                                                 │
│ Ctrl+C - Quit application                                                                        │
│ Ctrl+J - New line (Alt+Enter works for certain linux distros)                                    │
│ Ctrl+L - Clear the screen                                                                        │
│ Ctrl+X - Open input in external editor                                                           │
│ Ctrl+Y - Toggle YOLO mode                                                                        │
│ Enter - Send message                                                                             │
│ Esc - Cancel operation / Clear input (double press)                                              │
│ Shift+Tab - Toggle auto-accepting edits                                                          │
│ Up/Down - Cycle through your prompt history                                                      │
│                                                                                                  │
│ For a full list of shortcuts, see docs/keyboard-shortcuts.md                                     │
│                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
"
`;
